controller/controller.go

```diff
@@ -1516,9 +1516,6 @@ func (ctrl *ProvisionController) provisionClaimOperation(ctx context.Context, cl
 	if err := ctrl.volumeStore.StoreVolume(logger, claim, volume); err != nil {
 		return ProvisioningFinished, err
 	}
-	if err = ctrl.volumes.Add(volume); err != nil {
```

Comment:
We use this ugly hack (yes, it's ugly) to [prevent re-processing the same PVC again](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/a0dd6f07d1514e605869ec48ef6928e1133f2a04/controller/controller.go#L1422-L1427).

Consider a situation when provisioning succeeds, PV is saved to the API server, but that information has not been synced to the provisioner informers yet.

`syncClaim` gets called again for the same PVC from whatever reason, e.g. because of an added finalizer. The provisioner has no information that a volume was already provisioned for this claim. It will call the CSI driver again (that can be expensive), get the same volume as before (hopefully, some CSI drivers are not good at idempotency), and save a PV for it to etcd (again, expensive). That will fail, because the PV already exists. Nothing _bad_ actually happens, but it generates a lot useless load on the CSI driver + the API server. Our initial testing of the provisoner has shown that it dramatically reduces performance when many PVCs need to provisioned at the same time.

We would need a solution that:
1. Does not need a new API call [here](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go#L1422-L1427). Checking if the PV exists is simple code-wise, but it's expensive on the API server.
2. Skips volumes that were already provisioned.
3. Does not dramatically increase memory costs. Remembering PVCs provisioned in the last few minutes could be OK, remembering *all* of them since the provisioner startup half a year ago could be bad.
4. Is easily understandable and readable.

Reply:
@jsafrane  Thank you for your suggestion. I think this is an issue that occurs with the current informer list-watch mechanism. Having a small probability of encountering repeated handling in the CSI driver is not a problem; if a duplicate creation happens, it will result in an "already exists" error. I'm wondering if it's necessary to solve this with redundant code caching. If it is indeed necessary, I think it might be better to implement this in the client-go.

Reply:
The problem is that it's not a "small probability of encountering repeated handling", a vast majority of CreateVolume requests was repeated last time I checked.

Reply:
I'm quite curious about your testing approach. Could you share it for reference? Here's the test I conducted on my modified version: I added a logic to query the API server directly when determining that the volume does not exist in the cache. If the volume exists, it logs a message. After stress testing, I found that this occurs at a ratio of roughly 1/1000, which I think is relatively low. What do you think? Alternatively, if necessary, we could add a logic to query the API server after the cache check. What are your thoughts? @jsafrane